nomad/leader.go

```diff
@@ -59,37 +59,60 @@ var defaultSchedulerConfig = &structs.SchedulerConfiguration{
 func (s *Server) monitorLeadership() {
 	var weAreLeaderCh chan struct{}
 	var leaderLoop sync.WaitGroup
-	for {
-		select {
-		case isLeader := <-s.leaderCh:
-			switch {
-			case isLeader:
-				if weAreLeaderCh != nil {
-					s.logger.Error("attempted to start the leader loop while running")
-					continue
-				}
 
-				weAreLeaderCh = make(chan struct{})
-				leaderLoop.Add(1)
-				go func(ch chan struct{}) {
-					defer leaderLoop.Done()
-					s.leaderLoop(ch)
-				}(weAreLeaderCh)
-				s.logger.Info("cluster leadership acquired")
-
-			default:
-				if weAreLeaderCh == nil {
-					s.logger.Error("attempted to stop the leader loop while not running")
-					continue
-				}
+	leaderStep := func(isLeader bool) {
+		switch {
+		case isLeader:
+			if weAreLeaderCh != nil {
+				s.logger.Error("attempted to start the leader loop while running")
+				return
+			}
 
-				s.logger.Debug("shutting down leader loop")
-				close(weAreLeaderCh)
-				leaderLoop.Wait()
-				weAreLeaderCh = nil
-				s.logger.Info("cluster leadership lost")
+			weAreLeaderCh = make(chan struct{})
+			leaderLoop.Add(1)
+			go func(ch chan struct{}) {
+				defer leaderLoop.Done()
+				s.leaderLoop(ch)
+			}(weAreLeaderCh)
+			s.logger.Info("cluster leadership acquired")
+
+		default:
+			if weAreLeaderCh == nil {
+				s.logger.Error("attempted to stop the leader loop while not running")
+				return
 			}
 
+			s.logger.Debug("shutting down leader loop")
+			close(weAreLeaderCh)
+			leaderLoop.Wait()
+			weAreLeaderCh = nil
+			s.logger.Info("cluster leadership lost")
+		}
+	}
+
+	wasLeader := false
+	for {
+		select {
+		case isLeader := <-s.leaderCh:
+			if wasLeader != isLeader {
+				wasLeader = isLeader
+				// normal case where we went through a transition
+				leaderStep(isLeader)
+			} else if wasLeader && isLeader {
+				// Server lost but then gained leadership immediately.
+				// During this time, this server may have received
+				// Raft transitions that haven't been applied to the FSM
+				// yet.
+				// Ensure that that FSM caught up and eval queues are refreshed
+				s.logger.Error("cluster leadership flapped, lost and gained leadership immediately.  Leadership flaps indicate a cluster wide problems (e.g. networking).")
+
+				leaderStep(false)
+				leaderStep(true)
+			} else {
+				// Server gained but lost leadership immediately
+				// before it reacted; nothing to do, move on
+				s.logger.Error("cluster leadership flapped, gained and lost leadership immediately.  Leadership flaps indicate a cluster wide problems (e.g. networking).")
```

Comment:
I think this might be more accurate? I'm not sure gaining-and-losing-before-establishing leadership indicates a clusterwide problem.

```suggestion
				s.logger.Warn("cluster leadership gained and lost. Could indicate network issues, memory paging, or high CPU load.")
```

[AFAICT we only use the defaults for leadership election](https://github.com/hashicorp/nomad/blob/master/nomad/config.go#L342) (1s timeout causes an election, elections have 1s timeout) and don't expose them for customizing. It seems like clusters without demanding scheduler throughput or latency may prefer higher timeouts to reduce elections (and therefore flapping).

If we allowed Raft to be configurable we could at least point users toward those docs when these cases are encountered?

Reply:
re: configuration - agreed - we should consider porting consul's raft_multipler [1] and study their defaults.  In some Consul testing, they increased their default leadership timeout to 5 seconds to account for low powered instances (e.g. `t2.medium`).

[1] https://www.consul.io/docs/install/performance.html#minimum-server-requirements